source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -786,6 +857,25 @@ void Filter::onGenericPoolReady(StreamInfo::StreamInfo* info,
   }
   upstream_info.setUpstreamHost(host);
   upstream_info.setUpstreamSslConnection(ssl_info);
+
+  // If downstream is already closed, close the upstream connection immediately.
+  if (downstream_closed_) {
+    if (upstream_) {
+      Tcp::ConnectionPool::ConnectionData* conn_data =
+          upstream_->onDownstreamEvent(Network::ConnectionEvent::LocalClose);
+      if (conn_data != nullptr &&
+          conn_data->connection().state() != Network::Connection::State::Closed) {
+        config_->drainManager().add(config_->sharedConfig(),
+                                    Tcp::ConnectionPool::ConnectionDataPtr(conn_data),
+                                    std::move(upstream_callbacks_), std::move(idle_timer_),
+                                    idle_timeout_, read_callbacks_->upstreamHost());
+      }
+      upstream_.reset();
+      disableIdleTimer();
+    }
+    return;
+  }
```

Comment:
Why this is necessary? I didn't get it. Sorry.

Reply:
Yeah, it seems to be not needed anymore. I'll remove it. Thanks!